product_customer_code/models/product.py

```diff
@@ -0,0 +1,39 @@
+# -*- coding: utf-8 -*-
+
+###########################################################################
+#    Module Writen to OpenERP, Open Source Management Solution
+#
+#    Copyright (c) 2012 Vauxoo - http://www.vauxoo.com
+#    All Rights Reserved.
+#    info@vauxoo.com
+############################################################################
+#    Coded by: Rodo (rodo@vauxoo.com),Moy (moylop260@vauxoo.com)
+############################################################################
+
+from odoo import models, fields, api
+
+
+class ProductProduct(models.Model):
+    _inherit = "product.product"
+
+    product_customer_code_ids = fields.One2many(
+        comodel_name='product.customer.code',
+        inverse_name='product_id',
+        string='Customer Codes',
+        copy=False
+    )
+
+    @api.model
+    def name_search(self, name='', args=None, operator='ilike', limit=100):
+        res = super(ProductProduct, self).name_search(
+            name=name, args=args, operator=operator, limit=limit)
+        if not res:
+            partner_id = self._context.get('partner_id')
+            if partner_id:
+                product_customer_code_obj = self.env['product.customer.code']
+                product_codes = product_customer_code_obj.search([
+                    ('product_code', '=', name),
```

Comment:
I do agree that showing results based on both product name / default_code and product_code would be more flexible. Otherwise the scenario of "search by code only when by name / default_code fails" is in effect.

If by some chance the name/default_code of a product is the same with a product_code you cannot select it.

Searching by child_of the commercial_partner_id is not a bad idea.

Maybe someone else can pitch in to this and offer an opinion